packages/ckb-indexer/src/ckbIndexerFilter.ts

```diff
@@ -0,0 +1,351 @@
+import {
+  Cell,
+  DataWithSearchMode,
+  HexadecimalRange,
+  Script,
+  ScriptWrapper,
+  blockchain,
+} from "@ckb-lumos/base";
+import { CKBIndexerQueryOptions, SearchKey } from "./type";
+import { bytes } from "@ckb-lumos/codec";
+import { BI } from "@ckb-lumos/bi";
+
+function convertQueryOptionToSearchKey(
+  queryOptions: CKBIndexerQueryOptions
+): SearchKey {
+  let searchKeyLock: Script | undefined;
+  let searchKeyType: Script | undefined;
+  let searchKey: Required<SearchKey>;
+
+  const queryLock = queryOptions.lock;
+  const queryType = queryOptions.type;
+  if (queryLock) {
+    if (instanceOfScriptWrapper(queryLock)) {
+      searchKeyLock = queryLock.script;
+    } else {
+      searchKeyLock = queryLock;
+    }
+  }
+  if (queryType && queryType !== "empty") {
+    if (instanceOfScriptWrapper(queryType)) {
+      searchKeyType = queryType.script;
+    } else {
+      searchKeyType = queryType;
+    }
+  }
+
+  if (searchKeyLock) {
+    searchKey = {
+      script: searchKeyLock,
+      scriptType: "lock",
+      scriptSearchMode: instanceOfScriptWrapper(queryLock)
+        ? queryLock.searchMode || "prefix"
+        : "prefix",
+      filter: {},
+    };
+    searchKeyType && (searchKey.filter.script = searchKeyType);
+  } else if (searchKeyType) {
+    searchKey = {
+      script: searchKeyType,
+      scriptType: "type",
+      scriptSearchMode: instanceOfScriptWrapper(queryType)
+        ? queryType.searchMode || "prefix"
+        : "prefix",
+      filter: {},
+    };
+  } else {
+    throw new Error("query.lock and query.type can't be both empty");
+  }
+
+  const {
+    fromBlock,
+    toBlock,
+    outputDataLenRange,
+    outputCapacityRange,
+    scriptLenRange,
+  } = queryOptions;
+
+  if (fromBlock || toBlock) {
+    searchKey.filter.blockRange = [
+      fromBlock || "0x0",
+      toBlock
+        ? BI.from(toBlock).add(1).toHexString()
+        : "0x" + Number.MAX_SAFE_INTEGER.toString(16),
+    ];
+  }
+
+  if (outputDataLenRange) {
+    searchKey.filter.outputDataLenRange = [
+      outputDataLenRange[0],
+      BI.from(outputDataLenRange[1]).add(1).toHexString(),
+    ];
+  }
+
+  if (outputCapacityRange) {
+    searchKey.filter.outputCapacityRange = [
+      outputCapacityRange[0],
+      BI.from(outputCapacityRange[1]).add(1).toHexString(),
+    ];
+  }
+
+  if (scriptLenRange) {
+    searchKey.filter.scriptLenRange = [
+      scriptLenRange[0],
+      BI.from(scriptLenRange[1]).add(1).toHexString(),
+    ];
+  }
+
+  if (queryType === "empty") {
+    searchKey.filter.scriptLenRange = ["0x0", "0x1"];
+  }
+
+  return searchKey;
+}
+
+function filterByQueryOptions(
```

Comment:
I think we should realize that the `SearchKey` is for search/filter/match, and the `order` and `limit` is not for searching

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/7511174/236479407-051d4e8d-ed65-494e-9ae8-31aaedf146c6.png">

This method is used to filter the cells that match the conditions, it should not be mixed with the sorting limit and other operations. We can think of how we would use it 

```ts
async function *collectWithPendingCells(): AsyncGenerator<Cell> {
  for await (let cell of indexer.get(convertToSearchKey(options, cursor)) {
    if (isMatchQueryOption(cell)) yield cell; 
  }

  for (let cell of pendingCells) {
    if(isMatchQueryOption(cell)) yield cell
  }
}
```

Reply:
updated